File: 090 Subsets II/subsets2.cs
public class Solution {	// Author: Xinyu Chen
    public IList<IList<int>> SubsetsWithDup(int[] nums) {
        var result = new List<IList<int>>();
        result.Add(new List<int>());
        Array.Sort(nums);
        int i = 0;
        while (i < nums.Length)
        {
            int dup = 0;
            while (i + dup < nums.Length && nums[i] == nums[i + dup]) dup++;
            int curr = result.Count;
            for (int j = 0; j < curr; j++)
            {
                var row = new List<int>(result[j]);
                for (int k = 0; k < dup; k++)
                {
                    row.Add(nums[i]);
                    var r = new List<int>(row);
                    result.Add(r);
                }
            }
            i += dup;
        }
        return result;
    }
}

File: 090 Subsets II/subsets2.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        len = nums.size();
        sort(nums.begin(), nums.end());
        vector<int> entry;
        helper(nums, entry, 0);
        return result;
    }
    
    void helper(vector<int>& nums, vector<int>& entry, int curr) {
        result.push_back(entry);
        for (int i = curr; i < len; i++) {
            entry.push_back(nums[i]);
            helper(nums, entry, i + 1);
            entry.pop_back();
            while (i < len - 1 && nums[i] == nums[i + 1]) i++;
        }
    }
    
private:
    vector<vector<int>> result;
    int len;
}; // Author: XC
